### PR TITLE
Assignment1

### PR DESCRIPTION
**Pull request by ionastasi**
Added solutions for tasks.

### CODE REVIEW

File: assignment1/isPermutation.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+def isPermutation(str1, str2):
```
Comment (flerdacodeu) [inline assignment1/isPermutation.py:1]: Add a docstring describing what the function does, what its arguments are, and what it returns.

File: assignment1/isPermutation.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    if str1 is None or str2 is None:
+        return False
+    str1 = str1.lower()
+    str2 = str2.lower()
+    if str1 == str2:
+        return True
+    allSymbols = dict()
+    for symb in str1:
```
Comment (flerdacodeu) [inline assignment1/isPermutation.py]: I think using `symbol` is not much longer, but clearer.

File: assignment1/isPermutation.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+def isPermutation(str1, str2):
+    if str1 is None or str2 is None:
+        return False
+    str1 = str1.lower()
+    str2 = str2.lower()
+    if str1 == str2:
+        return True
+    allSymbols = dict()
```
Comment (flerdacodeu) [inline assignment1/isPermutation.py]: The meaning of `allSymbols` is not clear. It is a dictionary, but what are the keys and what are the values?
You can either find self-describing name or add a comment above this line to describe what it is.

File: assignment1/isPermutation.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        allSymbols[symb] -= 1
+    for symb in allSymbols:
+        if allSymbols[symb]:
+            return False
+    return True
+
+def _test():
+    while True:
+        print("\ninput two string")
+        s1 = input()
```
Comment (flerdacodeu) [inline assignment1/isPermutation.py]: You can also use the argument to `input()` to print a useful prompt.
Something like:

```
Input to strings:
> abc
> cba
the two strings are permutations of each other.
```

File: assignment1/isPermutation.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        s2 = input()
+        print(isPermutation(s1, s2))
+
+if __name__ == "__main__":
+    _test()
```
Comment (flerdacodeu) [inline assignment1/isPermutation.py]: You actually want to write also automated tests for this code.
Using the `unittest` module you want to check some common and corner cases, so that we can make sure it works correctly.

File: assignment1/kthInLinkedList.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+class Node:
```
Comment (flerdacodeu) [inline assignment1/kthInLinkedList.py]: Add docstring for these classes.

File: assignment1/kthInLinkedList.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    def __init__(self, _key = 0, _next = None):
+        self.key = _key
+        self.next = _next
+
+    def __str__(self):
+        return str(self.key)
+
+def kthInLinkedList(linkedList, k):
```
Comment (flerdacodeu) [inline assignment1/kthInLinkedList.py]: Add a docstring for this function.

**Note**: the current implementation returns a `Node` but you probably want to return a value instead.

File: assignment1/kthInLinkedList.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    if k < 0:
+        return None
+    ptr1 = ptr2 = linkedList
+    cnt = -1
+    while cnt != k:
```
Comment (flerdacodeu) [inline assignment1/kthInLinkedList.py]: It is not immediately clear what this code is doing.
Maybe adding a couple of comments would help here.

**Review by flerdacodeu — CHANGES_REQUESTED** [review]
A few suggestions for you.

If you need any help (e.g., about how to get unit tests set up), feel free to reach out to me and I can give you examples and details.

File: assignment1/isPermutation.py (hunk 1 above)
Comment (ionastasi) [reply assignment1/isPermutation.py:1]: Done.

File: assignment1/isPermutation.py (hunk 2 above)
Comment (ionastasi) [reply assignment1/isPermutation.py]: Done.

File: assignment1/isPermutation.py (hunk 3 above)
Comment (ionastasi) [reply assignment1/isPermutation.py]: Done.

File: assignment1/isPermutation.py (hunk 4 above)
Comment (ionastasi) [reply assignment1/isPermutation.py]: Done.

File: assignment1/isPermutation.py (hunk 5 above)
Comment (ionastasi) [reply assignment1/isPermutation.py]: Done.

File: assignment1/kthInLinkedList.py (hunk 6 above)
Comment (ionastasi) [reply assignment1/kthInLinkedList.py]: Done.

File: assignment1/kthInLinkedList.py (hunk 7 above)
Comment (ionastasi) [reply assignment1/kthInLinkedList.py]: Done.

File: assignment1/kthInLinkedList.py (hunk 8 above)
Comment (ionastasi) [reply assignment1/kthInLinkedList.py]: Done.

**Comment by ionastasi** [comment]
@flerdacodeu I fixed all your comments, can you please look again?

File: assignment1/isPermutation.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+def isPermutation(str1, str2):
+    """Function that checks if one string is a permutation of the other.
+
+    Args:
+        str1: fist string;
```
Comment (flerdacodeu) [inline assignment1/isPermutation.py]: nit: fist -> first

File: assignment1/isPermutation.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        a bool, True if the first string is a permutation of the other,
+        and False otherwise.
+    """
+    if str1 is None or str2 is None:
+        return False
+    str1 = str1.lower()
+    str2 = str2.lower()
+    if str1 == str2:
+        return True
+    sybsFromStr1 = dict()
```
Comment (flerdacodeu) [inline assignment1/isPermutation.py]: I would use `symbolsFromStr1` to make it clearer. Abbreviating the name makes it harder when you have to read the same code one year later or someone else is trying to understand it.

File: assignment1/isPermutation.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    for symbol in str1:
+        sybsFromStr1[symbol] = sybsFromStr1.get(symbol, 0) + 1
+    for symbol in str2:
+        if symbol not in sybsFromStr1.keys():
```
Comment (flerdacodeu) [inline assignment1/isPermutation.py]: If `d` is a dictionary:
```python
s in d.keys()
```
and
```python
s in d
```
have the same meaning, i.e., checking if the symbol is in the keys is the same as checking if the symbol is in the dictionary.

You can simplify the `if`-statement here.

File: assignment1/isPermutation_test.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
+"""Tests for isPermutation()."""
+
+from isPermutation import *
```
Comment (flerdacodeu) [inline assignment1/isPermutation_test.py]: Usually, one wants to avoid importing `*` because it may import more things than you originally intended when someone adds more things to the module. In this case, you really only need the `isPermutation` function.

File: assignment1/kthInLinkedList.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        self.next = _next
+
+    def __str__(self):
+        return str(self.key)
+
+def kthInLinkedList(linkedList, k):
+    """Function that finds the kth to last element of a singly linked list.
+
+    Args:
+        linledList: a Node, the first element in the linked list;
```
Comment (flerdacodeu) [inline assignment1/kthInLinkedList.py]: nit: linledList -> linkedList

File: assignment1/kthInLinkedList.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+        next: a Node, the next element in the linked list.
+    """
+    def __init__(self, _key = 0, _next = None):
+        self.key = _key
+        self.next = _next
+
+    def __str__(self):
+        return str(self.key)
+
+def kthInLinkedList(linkedList, k):
```
Comment (flerdacodeu) [inline assignment1/kthInLinkedList.py]: The name of the function does seem misleading. Call it `kthToLastInLinkedList` instead.

File: assignment1/kthInLinkedList_test.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
+"""Tests for kthInLinkedList()."""
+
+from kthInLinkedList import *
+import unittest
+
+class kthInLinkedListTest(unittest.TestCase):
+    def setUp(self):
+        self.n = 15
+        self.head = Node(self.n - 1)
```
Comment (flerdacodeu) [inline assignment1/kthInLinkedList_test.py]: It may be worth either having a comment to explain you are creating a list with values [14 13 12 10 .. 0] or do this using a private method on the test class (and make the name be enough to understand, e.g., `self.head = createReverseSequenceOfLength(15)`).

**Review by flerdacodeu — APPROVED** [review]
Thanks Anastasia! Well done.
Just a few minor comments.

Feel free to address them when you get a chance and then merge the pull request.

File: assignment1/isPermutation_test.py (hunk 12 above)
Comment (ionastasi) [reply assignment1/isPermutation_test.py]: Done

File: assignment1/isPermutation.py (hunk 10 above)
Comment (ionastasi) [reply assignment1/isPermutation.py]: Done

File: assignment1/isPermutation.py (hunk 9 above)
Comment (ionastasi) [reply assignment1/isPermutation.py]: Done

File: assignment1/isPermutation.py (hunk 11 above)
Comment (ionastasi) [reply assignment1/isPermutation.py]: Done

File: assignment1/kthInLinkedList.py (hunk 14 above)
Comment (ionastasi) [reply assignment1/kthInLinkedList.py]: Done

File: assignment1/kthInLinkedList.py (hunk 13 above)
Comment (ionastasi) [reply assignment1/kthInLinkedList.py]: Done

File: assignment1/kthInLinkedList_test.py (hunk 15 above)
Comment (ionastasi) [reply assignment1/kthInLinkedList_test.py]: I don't sure how to make private functions in the Python. As i know, `_` in the beginning of a function sname works smth like this. But I don't sure, may be it can be made by smth else.

**Comment by ionastasi** [comment]
@flerdacodeu I have a question for you under your last comment.

File: assignment1/kthInLinkedList_test.py (hunk 15 above)
Comment (flerdacodeu) [reply assignment1/kthInLinkedList_test.py]: That's correct.

In Python there is no modifier to specify a function or method is private.

The way this is achieve is by adding one or more `_` in front of the name.

For a module function, adding `_` in front of the name, e.g.:

```python
def _SortByDate(values):
  """Sorts the string values as if they were dates in the form DD/MM/YYYY.

  Args:
    values: a list of strings, each string representing a date.

  Returns:
    a list of strings, sorted in date order.
  """
  [...]
```
makes the function private, and, while it can be used as part of the implementation of the module, it should not be used outside that context.

For class member functions and variables, the rule are a little bit different.

A single `_` is used to identify members that are meant to be used by sub-classes but no other code outside the class itself. A double `_` is used to identify members that are meant to be private, so used neither by sub-classes nor by other code outside the class itself.

There are also some special member functions, like `__init__` to initialize an object or `__str__` to implement the to-string conversion, that are not considered private member functions; the names of these both start and end with `__`.

In this case, you probably want to define a private function at the top-level in the module named `_createReverseSequenceOfLength`.

**Review by flerdacodeu — APPROVED** [review]
LGTM
Thanks for updating the change.

Replied to your question. If you have any additional questions, please do let me know.

**Comment by ionastasi** [comment]
Okey, thanks!
